### PR TITLE
add ratelimit flag

### LINKED ISSUE #122: Add RateLimit flag
## Proposed changes
we can consider adding new flag type in goflags for ratelimit which accepts ratelimit along with unit

```
type RateLimit struct{
   MaxTokens unit
   Duration    time.Duration
}

rls := []RateLimit{}
flag.RateLimitVar(&rls, "-rls",nil,"rate limit along with unit ex:  10/ms , 30/m, 8/s ")
```
ref: https://github.com/projectdiscovery/subfinder/pull/891#pullrequestreview-1525004982

### PR DESCRIPTION
**Pull request by dogancanbakir**
Closes #122.

### CODE REVIEW

File: ratelimit_var.go [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+func (flagSet *FlagSet) RateLimitMapVar(field *RateLimitMap, long string, defaultValue []string, usage string) *FlagData {
+	return flagSet.RateLimitMapVarP(field, long, "", defaultValue, usage)
+}
+
+// RateLimitMapVarP adds a ratelimit flag with a short name and long name.
+// It is equivalent to RateLimitMapVar, and also allows specifying ratelimits in days (e.g., "hackertarget=2/d" 2 requests per day, which is equivalent to 24h).
+func (flagSet *FlagSet) RateLimitMapVarP(field *RateLimitMap, long, short string, defaultValue []string, usage string) *FlagData {
+	if field == nil {
+		panic(fmt.Errorf("field cannot be nil for flag -%v", long))
+	}
```
Comment (tarunKoyalwar) [inline ratelimit_var.go:89]: I think we should allow csv values in flag value (similar to how we normalize and use in goflags.StringSlice)
currently flag only supports this

```console
./basic -rls "h1=8/2h" -rls "h2=5/10s"
{"h2"="1m50s"="h1"="12h0m0s"}
map[h1:{8 12h0m0s} h2:{5 1m50s}]
```

```console
$ ./basic -rls "h1=8/2h,h2=5/10s"
invalid value "h1=8/2h,h2=5/10s" for flag -rls: parse error: time: unknown unit "h,h" in duration "12h,h2=5"
```

File: ratelimit_var.go [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+
+	var items string
+	for k, v := range rateLimitMap.kv {
+		items += fmt.Sprintf("\"%s\"=\"%s\"%s", k, v.Duration.String(), kvSep)
+	}
+	defaultBuilder.WriteString(stringsutil.TrimSuffixAny(items, ",", "="))
+	defaultBuilder.WriteString("}")
+	return defaultBuilder.String()
+}
+
```
Comment (tarunKoyalwar) [inline ratelimit_var.go:78]: not sure if this method is required but if required i think we can just json encode map
current implementation does not seem to be accurate

```console
./basic -rls "h1=8/2h"
{"h1"="12h0m0s"}
map[h1:{8 12h0m0s}]
```

**Review by tarunKoyalwar — CHANGES_REQUESTED** [review]
implementation LGTM !

```console
$ ./basic -rls "h1=8/s"                                             
{"h1"="1s"}

```

suggesting some minor changes

File: ratelimit_var.go (hunk 2 above)
Comment (dogancanbakir) [reply ratelimit_var.go]: This is required to implement flag.Value interface, so I modified it.

**Review by tarunKoyalwar — APPROVED** [review]
lgtm !

```console
$ go run . -rls "hackerone.com=8/1s,chaos.projectdiscovery.io=60/m"
Got RateLimits: {"hackerone.com":"8/1s","chaos.projectdiscovery.io":"60/1m0s"}
```

- just realised that time unit validation can be bypassed (ex: 1000m) so replaced with MaxRateLimitTime
- adds example usage in examples/basic/main.go file

**Review by Mzack9999 — APPROVED** [review]
lgtm!

We will need to find an easy way for the app using goflags to specify also a list of accepted keys that we will use for validation at runtime (similarly to enum) to avoid user typos (especially for long names). Maybe shall we track this in a follow up ticket?

**Comment by dogancanbakir** [comment]
@Mzack9999 Make sense. I opened an issue to track it https://github.com/projectdiscovery/goflags/issues/125. Feel free to edit/update.

**Pull request merged**